File: data_structures/go/stack.go
package main

type Stack interface {
	Push(value int) []int
	Pop() ([]int, int)
	Peek() int
}

type stack struct {
	stack []int
}

func NewStack() Stack {
	return &stack{}
}

func (s *stack) Push(value int) []int {
	return append(s.stack, value)
}

func (s *stack) Pop() ([]int, int) {
	return s.stack[:len(s.stack)-1], s.stack[len(s.stack)-1]
}

func (s *stack) Peek() int {
	return s.stack[len(s.stack)-1]
}

func (s *stack) isEmpty() bool {
	return len(s.stack) == 0
}

File: data_structures/java/Stack.java
package main;

public class Stack {
  public Node head;
  public Node tail;

  public Stack() {
    head = null;
    tail = null;
  }

  public void push(int data) {
    Node node = new Node(data);
    if (head == null) {
      head = node;
      tail = node;
    } else {
      node.next = head;
      head = node;
    }
  }

  public int pop() {
    int data = head.data;
    head = head.next;
    return data;
  }

  public int peek() {
    return head.data;
  }

  public boolean isEmpty() {
    return head == null;
  }
}
